### PR TITLE
Make all MultiMap works the same way

### PR DESCRIPTION
**Pull request by testn**
CaseInsensitiveHeaders/HeadersAdapter/Http2HeadersAdapter seem to work differently as the contract of MultiMap interface is not quite clear. I took that CaseInsensitiveHeaders is supposed to be the most reliable one.

BTW, I actually prefer the behavior of MultiMap interface to match DefaultHeaders in Netty especially when there are multiple entries per one header name.

### CODE REVIEW

**Comment by vietj** [comment]
can you explain what you mean by what you prefer ?

**Comment by testn** [comment]
i.e. size() returns entries.size() rather than names.size()

**Comment by vietj** [comment]
MultiMap#size() returns the number of keys per javadoc.

That would anyway break compatiblity.

If you need such value, then you can use `entries().size()` instead.

File: src/main/java/io/vertx/core/http/impl/HeadersAdaptor.java [hunk 1]
```diff
@@ -127,7 +136,44 @@ public MultiMap clear() {
 
   @Override
   public Iterator<Map.Entry<String, String>> iterator() {
-    return headers.iterator();
+    return new Iterator<Map.Entry<String, String>>() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/HeadersAdaptor.java]: what's the point of wrapping ?

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 2]
```diff
@@ -95,10 +96,14 @@ public int size() {
...
   @Override
   public List<Map.Entry<String, String>> entries() {
-    return headers.names()
-        .stream()
-        .map(name -> new AbstractMap.SimpleEntry<>(name.toString(), headers.get(name).toString()))
-        .collect(Collectors.toList());
+    if (isEmpty()) {
+      return Collections.emptyList();
+    }
+    List<Map.Entry<String, String>> list = new ArrayList<Map.Entry<String, String>>(headers.size());
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: I would prefer to return a list wrapper instead of creating a new ArrayList, beside the returned object must be immutable.

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 3]
```diff
@@ -95,10 +96,14 @@ public int size() {
...
+    for (Map.Entry<String, String> entry : this) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: this changes the behavior I think, the previous was I think wrong, i.e it was not returning duplicate entries.

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 4]
```diff
@@ -113,29 +118,27 @@ public boolean isEmpty() {
...
-          };
-        }
-        @Override
-        public int size() {
-          return headers.size();
-        }
-      };
-    }
-    return names;
+    Set<CharSequence> nameSet = headers.names();
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: why did you remove the caching ?

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 5]
```diff
@@ -201,7 +204,44 @@ public MultiMap clear() {
 
   @Override
   public Iterator<Map.Entry<String, String>> iterator() {
-    return entries().iterator();
+    return new Iterator<Map.Entry<String, String>>() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: what's the point of wrapping this ?

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 6]
```diff
@@ -113,29 +118,27 @@ public boolean isEmpty() {
...
+          }
+          @Override
+          public String next() {
+            return it.next().toString();
+          }
+        };
+      }
+      @Override
+      public int size() {
+        return nameSet.size();
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: I think indeed this is more correct to return the `nameSet.size()`

**Comment by vietj** [comment]
I've done more review

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java [hunk 7]
```diff
@@ -95,10 +96,14 @@ public int size() {
...
-        .map(name -> new AbstractMap.SimpleEntry<>(name.toString(), headers.get(name).toString()))
-        .collect(Collectors.toList());
+    if (isEmpty()) {
+      return Collections.emptyList();
+    }
+    List<Map.Entry<String, String>> list = new ArrayList<Map.Entry<String, String>>(headers.size());
+    for (Map.Entry<String, String> entry : this) {
+      list.add(entry);
+    }
+    return list;
```
Comment (testn) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: since we do not cache the result, we don't need to make it immutable.

File: src/main/java/io/vertx/core/http/impl/HeadersAdaptor.java (hunk 1 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/HeadersAdaptor.java]: let's stick to simplicity and remove it

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java (hunk 7 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: the javadoc says it is immutable

Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: and the `Collections.empty()` is immutable as well

File: src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java (hunk 3 above)
Comment (testn) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: I think dup entries are required. Otherwise, ClientConnection.createRequest will not work correctly.

Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java]: yes it should list all entries - just thinking we should cache the instance
